torch/utils/_python_dispatch.py

```diff
@@ -2,33 +2,76 @@
 import contextlib
 from typing import Iterator
 
-# Context manager that causes all pytorch operators to dispatch to the passed-in
-# type's __torch_dispatch__ function.
-# operation that accepts no tensors but returns a tensor.
-#
-# enable_python_mode is affected by torch._C._DisableTorchDispatch.
-#
-# NB: Calling an operator inside __torch_dispatch__ does go through
-# __torch_dispatch__ again. Please use _DisableTorchDispatch inside
-# __torch_dispatch__ to prevent infinite recursion.
+from torch.utils._mode_utils import _enable_mode, _ModeInfo
+from torch._C import _get_python_mode, _set_python_mode
+from dataclasses import dataclass
+
+@dataclass
+class PythonModeInfo(_ModeInfo):
+    def __init__(self):
+        # hacky because torch_function mode and python_mode don't yet have parity
+        super().__init__(mode_name="python", mode_class=type(None),
+                         base_mode_class=type(None),
+                         required_fn="__torch_dispatch__")
+
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or isinstance(mode, type) and issubclass(mode, (torch.Tensor,))
```

Comment:
You're not going to require torch dispatch modes to inherit from a TorchDispatchMode superclass? (If you did, then it would have been possible to shared `is_allowed_type` across the two implementations.